File: local-resolver/bulk_list_format.py
"""Bulk import line formatting (mirrors src/bulkListFormat.js)."""

from __future__ import annotations

import re


def parse_bulk_line(line: str) -> dict[str, str] | None:
    trimmed = str(line or "").strip()
    if not trimmed:
        return None

    title_part = trimmed
    link = ""
    pipe_idx = trimmed.find("|")
    if pipe_idx >= 0:
        title_part = trimmed[:pipe_idx].strip()
        link = trimmed[pipe_idx + 1 :].strip()

    if re.match(r"^https?://", trimmed, re.I) and not title_part:
        return {"title": "", "artist": "", "link": trimmed}

    by_match = re.match(r"^(.+?)\s+by\s+(.+)$", title_part, re.I)
    if by_match:
        return {"title": by_match.group(1).strip(), "artist": by_match.group(2).strip(), "link": link}

    dash = re.match(r"^(.+?)\s*[—–-]\s*(.+)$", title_part)
    if dash:
        return {"title": dash.group(1).strip(), "artist": dash.group(2).strip(), "link": link}

    tab_parts = [p.strip() for p in title_part.split("\t") if p.strip()]
    if len(tab_parts) >= 2:
        return {
            "title": tab_parts[0],
            "artist": tab_parts[1],
            "link": link or (tab_parts[2] if len(tab_parts) > 2 else ""),
        }

    return {"title": title_part, "artist": "", "link": link}


def format_bulk_line(row: dict[str, str]) -> str:
    title = str(row.get("title") or "").strip()
    artist = str(row.get("artist") or "").strip()
    link = str(row.get("link") or "").strip()
    line = title
    if artist:
        line = f"{title} by {artist}"
    if link:
        line = f"{line} | {link}"
    return line.strip()


def normalize_bulk_text(text: str) -> list[str]:
    lines = re.split(r"\r?\n", str(text or ""))
    rows: list[str] = []
    for line in lines:
        parsed = parse_bulk_line(line)
        if not parsed or (not parsed.get("title") and not parsed.get("link")):
            continue
        rows.append(format_bulk_line(parsed))
    return rows

File: src/bulkListFormat.js
export function parseBulkLine(line) {
  const trimmed = String(line || '').trim();
  if (!trimmed) return null;

  let titlePart = trimmed;
  let link = '';

  const pipeIdx = trimmed.indexOf('|');
  if (pipeIdx >= 0) {
    titlePart = trimmed.slice(0, pipeIdx).trim();
    link = trimmed.slice(pipeIdx + 1).trim();
  }

  if (/^https?:\/\//i.test(trimmed) && !titlePart) {
    return { title: '', artist: '', link: trimmed };
  }

  const byMatch = titlePart.match(/^(.+?)\s+by\s+(.+)$/i);
  if (byMatch) {
    return { title: byMatch[1].trim(), artist: byMatch[2].trim(), link: link };
  }

  const dash = titlePart.match(/^(.+?)\s*[—–-]\s*(.+)$/);
  if (dash) {
    return { title: dash[1].trim(), artist: dash[2].trim(), link: link };
  }

  const tabParts = titlePart.split('\t').map(function(p) { return p.trim(); }).filter(Boolean);
  if (tabParts.length >= 2) {
    return {
      title: tabParts[0],
      artist: tabParts[1],
      link: link || tabParts[2] || '',
    };
  }

  return { title: titlePart, artist: '', link: link };
}

export function formatBulkLine(row) {
  const title = String(row.title || '').trim();
  const artist = String(row.artist || '').trim();
  const link = String(row.link || '').trim();
  let line = title;
  if (artist) line = title + ' by ' + artist;
  if (link) line = line + ' | ' + link;
  return line.trim();
}

export function normalizeBulkTextLocally(text) {
  const lines = String(text || '').split(/\r?\n/);
  const rows = [];
  lines.forEach(function(line) {
    const parsed = parseBulkLine(line);
    if (!parsed || (!parsed.title && !parsed.link)) return;
    rows.push(formatBulkLine(parsed));
  });
  return rows.join('\n');
}

export function bulkLinesToCandidates(lines, tunebook, book) {
  const candidates = [];
  lines.forEach(function(line) {
    const parsed = parseBulkLine(line);
    if (!parsed) return;
    const tune = {
      name: parsed.title || 'Untitled',
      composer: parsed.artist || '',
      links: parsed.link ? [{ link: parsed.link, title: '', startAt: '', endAt: '' }] : [],
      voices: { '1': { meta: '', notes: [] } },
      books: book ? [book] : [],
    };
    candidates.push({
      tune: tune,
      sourceKind: 'bulk-text',
      youtubeUrl: parsed.link && /youtube|youtu\.be/i.test(parsed.link) ? parsed.link : '',
    });
  });
  return candidates;
}

export function driveListTextToBulkLines(text) {
  const trimmed = String(text || '').trim();
  if (!trimmed) return '';

  const lines = trimmed.split(/\r?\n/).filter(function(l) { return l.trim(); });
  if (lines.length <= 1) return trimmed;

  const first = lines[0];
  if (first.indexOf(',') >= 0 && !first.match(/^X:/)) {
    const header = first.toLowerCase();
    const hasTitle = header.indexOf('title') >= 0 || header.indexOf('name') >= 0 || header.indexOf('song') >= 0;
    if (hasTitle) {
      const cols = first.split(',').map(function(c) { return c.trim().toLowerCase(); });
      const titleIdx = cols.findIndex(function(c) { return c.indexOf('title') >= 0 || c === 'name' || c.indexOf('song') >= 0; });
      const artistIdx = cols.findIndex(function(c) { return c.indexOf('artist') >= 0 || c.indexOf('composer') >= 0; });
      const linkIdx = cols.findIndex(function(c) { return c.indexOf('url') >= 0 || c.indexOf('link') >= 0 || c.indexOf('youtube') >= 0; });
      const rows = [];
      lines.slice(1).forEach(function(line) {
        const parts = line.split(',').map(function(c) { return c.trim(); });
        rows.push(formatBulkLine({
          title: titleIdx >= 0 ? parts[titleIdx] : parts[0],
          artist: artistIdx >= 0 ? parts[artistIdx] : '',
          link: linkIdx >= 0 ? parts[linkIdx] : '',
        }));
      });
      return rows.join('\n');
    }
  }

  return normalizeBulkTextLocally(trimmed);
}
